### PR TITLE
set/get functions for listmode recon

### PR DESCRIPTION
**Pull request by NikEfth**
Mainly creating interface with SIRF. 
Work in progress.

### CODE REVIEW

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeData.cxx [hunk 1]
```diff
@@ -101,6 +111,22 @@ set_input_data(const shared_ptr<ExamData> & arg)
     this->list_mode_data_sptr = dynamic_pointer_cast<ListModeData>(arg);
 }
 
+template <typename TargetT>
+void
+PoissonLogLikelihoodWithLinearModelForMeanAndListModeData<TargetT>::
+set_cache_path(const std::string _cache_path)
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeData.cxx:117]: Use `cache_path_v` variable name here for the argument

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 2]
```diff
@@ -91,7 +91,6 @@ set_defaults()
   cache_lm_file = false;
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:91]: I guess delete?

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 3]
```diff
@@ -108,11 +107,11 @@ initialise_keymap()
 
   this->parser.add_key("num_events_to_use",&this->num_events_to_use);
   this->parser.add_key("max cache size", &cache_size);
-  this->parser.add_key("cache path", &cache_path);
   this->parser.add_key("recompute cache", &recompute_cache);
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:110]: Why is this here, but other cache cache variables one level up?

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 4]
```diff
@@ -122,6 +121,22 @@ set_num_subsets(const int new_num_subsets)
   return this->num_subsets;
 }
 
+template <typename TargetT>
+void
+PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin<TargetT>::
+set_proj_matrix_bybin(const shared_ptr<ProjMatrixByBin>& arg)
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:127]:
```suggestion
set_proj_matrix(const shared_ptr<ProjMatrixByBin>& arg)
```
I think is ok to shorten. Otherwise needs by_bin

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 5]
```diff
@@ -122,6 +121,22 @@ set_num_subsets(const int new_num_subsets)
...
+{
+    this->PM_sptr = arg;
+}
+
+template<typename TargetT>
+void
+PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin<TargetT>::
+set_proj_data_info_sptr(const shared_ptr<ProjData>& arg)
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:135]: Not sure are all why we need this. Maybe when only using cache and no input?

If we do need it

```suggestion
set_proj_data_info(const ProjData& arg)
```
I'm not sure if we have set_proj_data_info* elsewhere. If so, follow us naming convention, else remove _sptr as it's an internal detail how it stores it.

Best to pass a red to ProjData as passing a shared pointer implies to the user we're going to do something with the pointer

**Review by KrisThielemans — COMMENTED** [review]
Mostly ok but I'm reviewing from my phone...

There are quite a lot of white space changes which make my life harder. Ideally you remove them

**Comment by KrisThielemans** [comment]
Let me know when this can be merged

File: src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeData.h [hunk 6]
```diff
@@ -126,6 +128,18 @@ public PoissonLogLikelihoodWithLinearModelForMean<TargetT>
    virtual void start_new_time_frame(const unsigned int new_frame_num);
 
    ParseAndCreateFrom<TargetT, ListModeData> target_parameter_parser;
+   //! This flag is true when cache_size is more than zero.
+   bool cache_lm_file;
+   //! On the first cached run, the cache will be written in the working path of the reconstruction.
```
[the remark this answers is not in the document]
Comment (NikEfth) [reply src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeData.h:146]: I don't understand this comment

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx (hunk 3 above)
Comment (NikEfth) [reply src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx]: I fixed this (and other similars)

File: src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeData.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeData.h:146]: sorry. first sentence has to be "not necessarily in working folder", as you can now set it.

the rest is about the lines below. From what you wrote, I'm not sure what's going on. If it's easy to fix the description, then please do it...

**Comment by NikEfth** [comment]
@KrisThielemans, I Just finished with your first round of comments. Both here and in SIRF (https://github.com/SyneRBI/SIRF/pull/1107 )

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx (hunk 5 above)
Comment (KrisThielemans) [reply src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx]: ok. I understand. It's quite ugly, but fine.

I suppose the better way would be to create our own listmode fileformat that has the projdatainfo/examinfo and your "cached" file. But that's far too much work to do now.

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 7]
```diff
@@ -122,13 +116,69 @@ set_num_subsets(const int new_num_subsets)
...
+    if(this->skip_lm_input_file)
+    {
+        std::cout << "Dummy LM file" << std::endl;
+        this->list_mode_data_sptr.reset(new ListModeData_dummy(
+                                            arg.get_exam_info_sptr(),
+                                            proj_data_info_sptr));
+        this->frame_defs = arg.get_exam_info_sptr()->get_time_frame_definitions();
+    }
+    max_ring_difference_num_to_process = this->proj_data_info_sptr->get_scanner_sptr()->
+            get_num_rings() - 1;
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:142]: This seems quite dangerous. As no time to do this properly, use `warning()` and add a `\todo` in the doxygen of this function.

File: src/include/stir/listmode/ListModeData.h [hunk 8]
```diff
@@ -238,6 +238,41 @@ class ListModeData : public ExamData
   shared_ptr<const ProjDataInfo> proj_data_info_sptr;
 };
 
+
+class ListModeData_dummy : public ListModeData
```
Comment (KrisThielemans) [inline src/include/stir/listmode/ListModeData.h:242]: I'm not sure why you need this, but if you do, can you create a minimal doxygen header, saying that this is a temporary class that should not be used

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx (hunk 5 above)
Comment (NikEfth) [reply src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx]: Yeah, that could be a good final stage for the cache file.

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 9]
```diff
@@ -122,13 +116,69 @@ set_num_subsets(const int new_num_subsets)
...
+
+
 template<typename TargetT>
 bool
 PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin<TargetT>::
 actual_subsets_are_approximately_balanced(std::string& warning_message) const
 {
     if (skip_balanced_subsets)
+    {
+        warning(boost::format("We presume balaned subsets!"));
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:179]:
```suggestion
        warning("We skip the check on balanced subsets and presume they are balanced!"));
```
although I don't see why we need this capability (can't take any appreciable time really?). fine to leave it for now

File: src/include/stir/listmode/ListModeData.h (hunk 8 above)
Comment (KrisThielemans) [reply src/include/stir/listmode/ListModeData.h:242]: Also, correct grammar in messages below "does not have"

**Review by KrisThielemans — CHANGES_REQUESTED** [review]

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx (hunk 9 above)
Comment (NikEfth) [reply src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:179]: I always find it very slow :D

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx (hunk 7 above)
Comment (NikEfth) [reply src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:142]: If you come up with a better suggestion on where to have this, I am happy to move it. 
I am not sure that we need it for the Summer school. But better have it and fix it later than not have it and forget that we need it. I will add the warning, though

File: src/include/stir/listmode/ListModeData.h (hunk 8 above)
Comment (NikEfth) [reply src/include/stir/listmode/ListModeData.h:242]: I can not figure a way around it without changing way too much. 
If I change get_input_data() to return ExamData() the set_up breaks. 
If I store the ExamInfo separately, I have to change the set_up() of all base classes. 

The only alternative I could think of was to set a listmode file of another format, but then I would have to replace everything manually.

Comment (NikEfth) [reply src/include/stir/listmode/ListModeData.h:242]: It is very tricky.

Comment (NikEfth) [reply src/include/stir/listmode/ListModeData.h:263]: I added a brief doxygen header.

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx [hunk 10]
```diff
@@ -122,13 +116,70 @@ set_num_subsets(const int new_num_subsets)
...
+
+
 template<typename TargetT>
 bool
 PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin<TargetT>::
 actual_subsets_are_approximately_balanced(std::string& warning_message) const
 {
     if (skip_balanced_subsets)
+    {
+        warning("We skip the check on balanced subsets and presume they are balanced!"));
```
Comment (KrisThielemans) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx:180]:
```suggestion
        warning("We skip the check on balanced subsets and presume they are balanced!");
```
sorry

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
I had a typo. sorry.

Can you move the dummy listmode class to its own header? We get a lot of warnings about no return value, and they get repeated wherever we include `ListmodeData.h`.

**Comment by NikEfth** [comment]
Done

**Comment by KrisThielemans** [comment]
sadly Appveyor fails due to all missing return statements, see e.g. https://ci.appveyor.com/project/KrisThielemans/stir/builds/43603985/job/1ratk1ru57hnd28c#L717

**Comment by KrisThielemans** [comment]
MacOS job fails due to unrelated #1036

**Comment by NikEfth** [comment]
I added some returns. Hopefully it is happy now.
